lib/sqlitex.ex

```diff
@@ -7,19 +7,47 @@ defmodule Sqlitex do
     :esqlite3.open(path)
   end
 
-  def query(db, query, opts \\ []) do
-    {:ok, statement} = :esqlite3.prepare(query, db)
-    types = :esqlite3.column_types(statement) |> Tuple.to_list
-    columns = :esqlite3.column_names(statement) |> Tuple.to_list
-    rows = :esqlite3.fetchall(statement)
-    into = Keyword.get(opts, :into, [])
-    Sqlitex.Row.from(types, columns, rows, into)
-  end
-
   def with_db(path, fun) do
     {:ok, db} = open(path)
     res = fun.(db)
     close(db)
     res
   end
+
+  def exec(db, sql) do
+    :esqlite3.exec(sql, db)
+  end
+
+  def query(db, sql, opts \\ []) do
+    {params, into} = query_options(opts)
+    {:ok, statement} = :esqlite3.prepare(sql, db)
+    :ok = :esqlite3.bind(statement, params)
+    types = :esqlite3.column_types(statement)
+    columns = :esqlite3.column_names(statement)
+    rows = :esqlite3.fetchall(statement)
+    return_rows_or_error(types, columns, rows, into)
+  end
+
+  defp query_options(opts) do
+    params = Keyword.get(opts, :bind, [])
+    into = Keyword.get(opts, :into, [])
+    {params, into}
+  end
+
+  defp return_rows_or_error(_, _, {:error, _} = error, _), do: error
+  defp return_rows_or_error({:error, kind} = error, columns, rows, into) do
+    case kind do
+      :no_columns -> return_rows_or_error({}, columns, rows, into)
```

Comment:
Good catch.

Unfortunately, I made these changes too long ago to remember what the query was, but I think I remember the reasoning behind it:  I issued a select statement that caused `sqlite3_column_count()` to return 0.  This returned a tuple `{:error, :no_columns}` for the column types.  When `Sqlitex.Row.build_row()` was called, the error tuple truncated the row output to only two columns because it interpreted the error value as a set of types of "error" and "no_columns".

Having said that, I spent a considerable amount of time looking for a query to cause this error without success.  It's possible that some of the other code changes obsoleted this error condition.  Next time I will write a test or more descriptive commit message than "handles queries without columns" :(

I'll revert these lines to the original code.  I'm using Sqlitex to write [an Ecto adapter for SQLite3](https://github.com/jazzyb/sqlite_ecto), so I'm sure that if this remains an issue, I'll run back into the bug eventually.  In which case I can file a new pull request with tests.